File: unisex.c
//
// NAME:    readers-writers.c
// AUTHOR:  Andrew Meijer V00805554

//  I use MinGW on Windows10 and compile with the following command:
//      gcc -pthread readers-writers.c

#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

sem_t ms;       //male lightswitch control
sem_t fs;       //female lightswitch control
sem_t e;        //emptiness of the restroom/washroom
sem_t gp;       //multiplex: 3 people maximum.
sem_t lp;       //multiplex: 4 people bathing together is too many. (3 is ok)
int mc = 0;     //counter of the number of male users in lav'
int fc = 0;     //counter of the number of female users in lav'

void* female(void* arg)
{
    sem_wait(&fs);
    fc = fc + 1;
    if(fc == 1){
        sem_wait(&e); // first in locks
    }
    sem_post(&fs);

    sem_wait(&lp);
    printf("A woman uses the bathroom.\n");
    sleep(2);
    sem_post(&lp);

    sem_wait(&fs);
    fc = fc - 1;
    if(fc == 0){
        sem_post(&e); // last out unlocks
    }
    sem_post(&fs);
}

void* male(void* arg)
{
    sem_wait(&ms);
    mc = mc + 1;
    if(mc == 1){
        sem_wait(&e); // first in locks
    }
    sem_post(&ms);

    sem_wait(&gp);
    printf("A man uses the bathroom.\n");
    sleep(1);
    sem_post(&gp);

    sem_wait(&ms);
    mc = mc - 1;
    if(mc == 0){
        sem_post(&e); // last out unlocks
    }
    sem_post(&ms);
}

int main()
{
    sem_init(&e, 0, 1);
    sem_init(&ms, 0, 1);
    sem_init(&fs, 0, 1);
    sem_init(&gp, 0, 3);
    sem_init(&lp, 0, 3);

    pthread_t m1,m2,m3,m4,f1,f2,f3,f4;
    pthread_create(&m1,NULL,male,NULL);
    pthread_create(&m2,NULL,male,NULL);
    pthread_create(&m3,NULL,male,NULL);
    pthread_create(&m4,NULL,male,NULL);
    pthread_create(&f1,NULL,female,NULL);
    pthread_create(&f2,NULL,female,NULL);
    pthread_create(&f3,NULL,female,NULL);
    pthread_create(&f4,NULL,female,NULL);

    pthread_join(m1, NULL);
    pthread_join(m2, NULL);
    pthread_join(m3, NULL);
    pthread_join(m4, NULL);
    pthread_join(f1, NULL);
    pthread_join(f2, NULL);
    pthread_join(f3, NULL);
    pthread_join(f4, NULL);

    sem_destroy(&e);
    sem_destroy(&ms);
    sem_destroy(&fs);
    sem_destroy(&gp);
    sem_destroy(&lp);
    printf("Exiting Main Thread.\n");
    return 0;
}

File: unisex.py
'''
NAME: producers-consumers.py
AUTHOR: Andrew Meijer V00805554
'''
#!/usr/bin/python3
import queue
import threading
import time
import os
import psutil

class female (threading.Thread):
    def run(self):
        womanSwitch.acquire()
        global womanCounter
        womanCounter = womanCounter + 1
        if womanCounter == 1:
            empty.acquire() # first in locks
        womanSwitch.release()

        ladyPlex.acquire()
        print("A woman uses the bathroom.")
        time.sleep(10)
        ladyPlex.release()

        womanSwitch.acquire()
        womanCounter = womanCounter - 1
        if womanCounter == 0:
            empty.release() # last out unlocks
        womanSwitch.release()

class male (threading.Thread):
    def run(self):
        manSwitch.acquire()
        global manCounter
        manCounter = manCounter + 1
        if manCounter == 1:
            empty.acquire() # first in locks
        manSwitch.release()

        gentPlex.acquire()
        print("A man uses the bathroom.")
        time.sleep(3)
        gentPlex.release()

        manSwitch.acquire()
        manCounter = manCounter - 1
        if manCounter == 0:
            empty.release() # last out unlocks
        manSwitch.release()

# Main
empty = threading.Lock()
manSwitch = threading.Lock()
manCounter = 0
womanSwitch = threading.Lock()
womanCounter = 0
gentPlex = threading.Semaphore(3)
ladyPlex = threading.Semaphore(3)

# Create the human beings that must deploy upon the lavatory, as threads.
m1 = male()
m2 = male()
m3 = male()
m4 = male()
w1 = female()
w2 = female()
w3 = female()
w4 = female()

# there are a few different ways to test this
# adding the sleep times (above) showcases the multiplex.
m1.start()
w1.start()
m2.start()
w2.start()
m3.start()
w3.start()
m4.start()
w4.start()
m1.join()
m2.join()
m3.join()
m4.join()
w1.join()
w2.join()
w3.join()
w4.join()
print("All users have completed their bathtimes.")
# memory usage report:
# nts this is not counted in the comprehensibility measurements
pid = os.getpid()
ps = psutil.Process(pid)
mr = ps.memory_info()
print("Bytes used: ", mr.rss)
